### PR TITLE
feat: implement user info retrieval in room page

### PR DESCRIPTION
**Pull request by dextertanyj**

### CODE REVIEW

File: frontend/src/pages/RoomPage.tsx [hunk 1]
```diff
@@ -37,35 +67,135 @@ export const RoomPage = () => {
...
     if (!socket || !roomId) {
       return;
     }
 
-    socket.emit(ROOM_EVENTS.JOIN, roomId);
-
-    socket.on(ROOM_EVENTS.PARTNER_LEAVE, () => {
-      setMessage("The other user has left the room.");
+    socket.on(ROOM_EVENTS.DISCONNECT, (reason: string) => {
+      if (reason === "io server disconnect") {
```
Comment (ypinhsuan) [inline frontend/src/pages/RoomPage.tsx:81]: This looks like a magic string. Maybe we can create a const for the disconnection reasons.

**Review by ypinhsuan — APPROVED** [review]
Thanks for cleaning up my code

File: frontend/src/pages/QueuePage.tsx [hunk 2]
```diff
@@ -83,21 +64,28 @@ export const WaitingPage = () => {
...
-      clearTimeout(pageTimeout);
+      setMessage("Match found. Joining...");
+      clearTimeout(timeout);
       setTimeout(() => navigate(`/room/${match.roomId}`), 1000);
     });
 
     socket.on(MATCH_EVENTS.EXISTING_MATCH, (roomId: string) => {
-      setMessage("Exsiting match found. Joining back...");
-      clearTimeout(pageTimeout);
+      setMessage("Exsiting match found. Rejoining...");
```
Comment (junlong4321) [inline frontend/src/pages/QueuePage.tsx:87]: Typo

File: frontend/src/constants/socket.io.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+export const SOCKET_IO_DISCONNECT_REASION = {
```
Comment (junlong4321) [inline frontend/src/constants/socket.io.ts:1]: Typo - REASON

File: frontend/src/pages/RoomPage.tsx [hunk 4]
```diff
@@ -1,30 +1,61 @@
...
 import { ROOM_EVENTS, ROOM_NAMESPACE } from "~shared/constants";
+import {
+  JoinedPayload,
+  PartnerDisconnectPayload,
+  PartnerLeavePayload,
+} from "~shared/types/api";
+
+type participant = {
+  userId: number;
+  name?: string;
```
Comment (junlong4321) [inline frontend/src/pages/RoomPage.tsx:25]: Is there a reason why name is an optional field? If there exist a user then i assume there should be a name?

File: frontend/src/pages/RoomPage.tsx [hunk 5]
```diff
@@ -1,30 +1,61 @@
...
   const { socket, connect } = useSocket();
-  const [leaveRoom, setIsLeaveRoom] = useState<boolean>(false);
-  const [message, setMessage] = useState<string>("");
+  const [self, setSelf] = useState<participant>({
+    // We know that if the page renders, user is not null.
+    userId: user?.userId || NaN,
+    name: user?.name || "",
+    isConnected: false,
+  });
+  const [partner, setPartner] = useState<participant | undefined>(undefined);
```
Comment (junlong4321) [inline frontend/src/pages/RoomPage.tsx:39]: Should we standardize whether to use undefined or null? Wen Ling's PR uses null in the SelectDifficulty page

File: frontend/src/pages/RoomPage.tsx [hunk 6]
```diff
@@ -37,35 +68,135 @@ export const RoomPage = () => {
...
+        variant: "error",
+      });
     });
 
-    socket.on(ROOM_EVENTS.PARTNER_DISCONNECT, () => {
-      setMessage(
-        "The other user has disconnected. Waiting for reconnection...",
-      );
+    socket.on(ROOM_EVENTS.PARTNER_LEAVE, ({ userId }: PartnerLeavePayload) => {
+      if (userId === user?.userId) {
```
Comment (junlong4321) [inline frontend/src/pages/RoomPage.tsx:93]: If the self-disconnection is caught by the disconnect event, do we still need this guard conditional here?

File: shared/src/types/api/index.ts [hunk 7]
```diff
@@ -1,5 +1,6 @@
 export { LoginReq, LoginRes, VerifyEmailReq } from "./auth.dto";
 export { MatchRes } from "./match.dto";
+export * from "./room.dto";
```
Comment (junlong4321) [inline shared/src/types/api/index.ts:3]: Can we avoid wildcard imports?

**Review by junlong4321 — APPROVED** [review]
Looks good overall. Just some clarifications that I'd like to make + Minor typos

File: frontend/src/pages/RoomPage.tsx (hunk 4 above)
Comment (dextertanyj) [reply frontend/src/pages/RoomPage.tsx:25]: The name is not sent by the websocket to us, we have to fetch it separately to retrieve the partner's name.

File: frontend/src/pages/RoomPage.tsx (hunk 6 above)
Comment (dextertanyj) [reply frontend/src/pages/RoomPage.tsx:93]: It's possible that the socket disconnects and reconnects immediately, and during the disconnection, the server fires the event to all participants including the user that caused the disconnection. But we should ignore such events if we caused it.

**Pull request merged**